src/bgw/job.c

```diff
@@ -1300,22 +1304,42 @@ ts_bgw_job_set_job_entrypoint_function_name(char *func_name)
 	job_entrypoint_function_name = func_name;
 }
 
+/*
+ * Run job and set next start.
+ *
+ * job: Job to run and update
+ * func: Function to execute for the job
+ * initial_runs: Limit on the number of runs to do
+ * next_interval: Interval to use when computing next start
+ * atomic: Should be executed as a single transaction.
+ * mark: Mark the start and end of the function execution in job_stats
+ */
 bool
 ts_bgw_job_run_and_set_next_start(BgwJob *job, job_main_func func, int64 initial_runs,
-								  Interval *next_interval)
+								  Interval *next_interval, bool atomic, bool mark)
 {
 	BgwJobStat *job_stat;
-	bool ret = func();
+	bool had_error;
 
-	/* Now update next_start. */
-	StartTransactionCommand();
+	if (atomic)
+		StartTransactionCommand();
+
+	if (mark)
+		ts_bgw_job_stat_mark_start(job->fd.id);
 
+	had_error = func();
```

Comment:
note:
* before this patch: `func()` was called outside the transaction
* with this patch it will be placed inside it

as `func()` is telemetry:
* it starts/commits a transaction

I don't think that will cause a problem; but was this change intentional?

Reply:
It was intentional, but it works even if we move it outside because we never run with `mark` and `atomic` set to true at the same time.

The reason is that you need a transaction to mark the job start and end, and that only exists when you run from a portal. Now, since you can pass use `mark` as true and `atomic` as true as well, it suggest that you can call it in a situation where you do not have a transaction and expect the call to work anyway. Since the mark functions require a transaction, not wrapping the functions in a transaction would fail in this case, so I wrapped it in a transaction.